File: articles/cascading-failure-and-contagion/fortran/cascade_threshold.f90
program cascade_threshold
  implicit none
  character(len=16), dimension(4) :: sectors
  real, dimension(4) :: capacity, load, threshold
  real :: load_ratio
  integer :: i

  sectors = (/ 'energy          ', 'water           ', 'telecom         ', 'health          ' /)
  capacity = (/ 100.0, 85.0, 90.0, 95.0 /)
  load = (/ 62.0, 70.0, 58.0, 82.0 /)
  threshold = (/ 0.75, 0.70, 0.72, 0.78 /)

  print *, 'Cascade capacity diagnostics'

  do i = 1, 4
    load_ratio = load(i) / capacity(i)
    if (load_ratio >= threshold(i)) then
      print *, trim(sectors(i)), ' load_ratio=', load_ratio, ' threshold=', threshold(i), ' status=failure risk'
    else
      print *, trim(sectors(i)), ' load_ratio=', load_ratio, ' threshold=', threshold(i), ' status=within threshold'
    end if
  end do
end program cascade_threshold

File: articles/cascading-failure-and-contagion/cpp/cascade_threshold.cpp
#include <iostream>
#include <string>
#include <vector>

struct Node {
    std::string sector;
    double capacity;
    double load;
    double threshold;
};

int main() {
    std::vector<Node> nodes = {
        {"energy", 100.0, 62.0, 0.75},
        {"water", 85.0, 70.0, 0.70},
        {"telecom", 90.0, 58.0, 0.72},
        {"health", 95.0, 82.0, 0.78}
    };

    std::cout << "Cascade capacity diagnostics\n";

    for (const auto& node : nodes) {
        double load_ratio = node.load / node.capacity;
        std::string status = load_ratio >= node.threshold ? "failure risk" : "within threshold";
        std::cout << node.sector << " load_ratio=" << load_ratio << " threshold=" << node.threshold << " status=" << status << "\n";
    }

    return 0;
}
